astropy/utils/metadata.py

```diff
@@ -414,3 +415,73 @@ def __set__(self, instance, value):
                     instance._meta = value
             else:
                 raise TypeError("meta attribute must be dict-like")
+
+
+class MetaAttribute:
+    """
+    Descriptor to define custom attribute which gets stored in the object
+    ``meta`` dict and can have a defined default.
+
+    This descriptor is intended to provide a convenient way to add attributes
+    to a subclass of a complex class such as ``Table`` or ``NDData``.
+
+    This requires that the object has an attribute ``meta`` which is a
+    dict-like object.  The value of the MetaAttribute will be stored in a
+    new dict meta['__attributes__'] that is created when required.
+
+    Classes that define MetaAttributes are encouraged to support initializing
+    the attributes via the class ``__init__``.  For example::
+
+        for attr in list(kwargs):
+            descr = getattr(self.__class__, attr, None)
+            if isinstance(descr, MetaAttribute):
+                setattr(self, attr, kwargs.pop(attr))
+
+    The name of a ``MetaAttribute`` cannot be the same as any of the following:
+
+    - Keyword argument in the owner class ``__init__``
+    - Method or attribute of the "parent class", where the parent class is
+      taken to be ``owner.__mro__[1]``.
+
+    :param default: default value
+
+    """
+    def __init__(self, default=None):
+        self.default = default
+
+    def __get__(self, instance, owner):
+        # When called without an instance, return self to allow access
+        # to descriptor attributes.
+        if instance is None:
+            return self
+
+        # Get the __attributes__ dict and create if not there already.
+        meta = instance.meta.setdefault('__attributes__', {})
+        try:
+            value = meta[self.name]
+        except KeyError:
+            if self.default is not None:
```

Comment:
I wondered whether we should return `deepcopy(self.default)` here independent of the value of `default`, i.e., never set the attribute? But now I realize that if `self.default = {}`, that would go rather badly wrong... (that example of course also is a good reason for the copy).  So, :+1: to the implementation, but perhaps add a comment?